src/sql/query.rs

```diff
@@ -1597,6 +1592,29 @@ fn plan_aggregate(ecx: &ExprContext, sql_func: &Function) -> Result<AggregateExp
             }
         }
     };
+    if let Some(filter) = &sql_func.filter {
+        let cond = plan_expr(&ecx.with_name("FILTER"), filter, Some(ScalarType::Bool))?;
+        let cond_typ = ecx.column_type(&cond);
+        if cond_typ.scalar_type != ScalarType::Bool && cond_typ.scalar_type != ScalarType::Unknown {
+            bail!(
+                "WHERE expression in FILTER must have boolean type, not {:?}",
+                cond_typ
+            );
+        }
+        let expr_typ = ecx.scalar_type(&expr);
+        if func == AggregateFunc::CountAll {
+            // `count(*)` doesn't filter out NULLs, so rewrite as `count(TRUE)`.
+            // The replacement below will result in `count(CASE WHEN cond THEN
+            // TRUE ELSE NULL)`, which has the desired semantics.
+            func = AggregateFunc::Count;
+            expr = ScalarExpr::literal_true();
```

Comment:
TIOLI: This comment and code reads to me as though `count(*) FILTER (..)` is meaningless, which obviously isn't the case.

Reply:
Reworked the comment to be more explanatory about the whole filter approach, and moved it up a bit, which I think should be more clear!

Reply:
![image](https://user-images.githubusercontent.com/277161/78315474-197eb800-752b-11ea-8b66-77c6befb6828.png)